File: chat-backend/src/main/java/com/smallchat/backend/chat/application/usecase/CreateChatUseCase.java
package com.smallchat.backend.chat.application.usecase;

import org.springframework.stereotype.Service;

import com.smallchat.backend.chat.application.inputport.CreateChatInputPort;
import com.smallchat.backend.chat.application.outputport.ChatOutputPort;
import com.smallchat.backend.chat.application.outputport.EventOutputPort;
import com.smallchat.backend.chat.application.outputport.MessageOutputPort;
import com.smallchat.backend.chat.domain.event.ChatJoined;
import com.smallchat.backend.chat.domain.model.Chat;
import com.smallchat.backend.chat.domain.model.vo.Message;
import com.smallchat.backend.chat.domain.model.vo.SystemMessage;
import com.smallchat.backend.chat.framework.web.dto.CreateChatDto;
import com.smallchat.backend.global.utils.TokenPayload;
import com.smallchat.backend.user.application.inputport.ValidateUserInputPort;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class CreateChatUseCase implements CreateChatInputPort {

    private final ChatOutputPort chatOutputPort;
    private final EventOutputPort eventOutputPort;
    private final MessageOutputPort messageOutputPort;

    //todo inputport 수정 필요
    private final ValidateUserInputPort validateUserInputPort;

    @Override
    public String createChat(TokenPayload tokenPayload, CreateChatDto.Request request) {
        String ownerId = tokenPayload.userId();
        validateUserInputPort.hasReachedMaxChatLimit(ownerId);
        Chat chat = Chat.of(ownerId, request.chatName());
        String chatId = chatOutputPort.save(chat).getChatId();
        messageOutputPort.save(Message.systemMessage(SystemMessage.CHAT_CREATED, chat.getName(), chatId));
        eventOutputPort.occurJoinChatEvent(new ChatJoined(ownerId, chatId));
        return chatId;
    }
}

File: chat-backend/src/main/java/com/smallchat/backend/chat/application/usecase/CreateChatUseCase.kt
package com.smallchat.backend.chat.application.usecase

import com.smallchat.backend.chat.domain.interfaces.ChatRepository
import com.smallchat.backend.chat.domain.interfaces.ChatUserRepository
import com.smallchat.backend.chat.domain.model.Chat
import com.smallchat.backend.chat.domain.model.ChatUser
import com.smallchat.backend.chat.domain.model.vo.ChatRole
import com.smallchat.backend.chat.domain.policy.ChatParticipationPolicy
import com.smallchat.backend.chat.interfaces.web.dto.CreateChatDto
import com.smallchat.backend.global.infrastructure.rabbitMq.ChatQueueManager
import com.smallchat.backend.global.utils.AuthenticatedUser
import lombok.RequiredArgsConstructor
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@RequiredArgsConstructor
class CreateChatUseCase(
    private val chatUserRepository: ChatUserRepository,
    private val chatRepository: ChatRepository,
    private val chatParticipationPolicy: ChatParticipationPolicy,
    private val chatQueueManager: ChatQueueManager
) {

    @Transactional
    fun createChat(authUser: AuthenticatedUser, request: CreateChatDto.Request): String {
        val userId = authUser.userId
        val chatName = request.chatName
        chatParticipationPolicy.ensureUserCanJoin(userId)
        val savedChat = chatRepository.save(Chat(name = chatName))
        chatUserRepository.save(ChatUser(null, savedChat, userId, ChatRole.HOST))
        chatQueueManager.createQueueForChat(savedChat.chatIdOrThrow)
        return savedChat.chatIdOrThrow
    }
}
